### DISCUSSION TITLE
Python Dependency Management

### DISCUSSION THREAD

**Discussion by tusharg1993** [apache/airflow#29966] [General] [answered]
Hello. How does Airflow maintain python package dependencies?

Say if my workflow task depends on numpy, do I have to make sure numpy is installed on all of the worker nodes as well?

Also, is there a way to prevent dependency conflicts between various workflows? Let's say if workflow1 depends on numpy v1.x and workflow2 depends on numpy ver2.x. Is there a way to isolate the python environment between the two workflows in same airflow installations?

**Comment by pierrejeambrun** [comment, marked as answer]
Hi @tusharg1993,

Yes, if your tasks depends on some library, it needs to be available in your worker node as well.

Airflow dependencies are handled in a special way as it is both an application and a library, you can read more about it here:
https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

There are multiple ways of dealing with this, everything there is to know about python and system dependencies management can be found here:
https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#handling-conflicting-complex-python-dependencies

Marking this as answered, don't hesitate to re-open if the doc doesn't answer all your questions.
